stage/stagein.man
.\" @(#)$RCSfile: stagein.man,v $ $Revision: 1.15 $ $Date: 2001/09/26 09:13:54 $ CERN IT-PDP/DM Jean-Philippe Baud
.\" Copyright (C) 1994-1999 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH STAGEIN l "$Date: 2001/09/26 09:13:54 $"
.SH NAME
stagein \- allocate space in a disk pool and copy files from tape to disk or
from disk to disk
.SH SYNOPSIS
.B stagein  
[
.BI -A " allocation_policy"
] [
.BI -b " max_block_size"
] [
.BI -c " off"
] [
.BI -C " charconv"
] [
.BI -d " density"
] [ 
.BI -E " error_action"
] [
.BI -F " record_format" 
] [
.BI -f " file_id"
] [
.BI -G 
] [ 
.BI -g " device_group_name"
] [
.BI -h " host"
] [
.BI -I " external_filename"
] [ 
.BI -L " record_length"
] [ 
.BI -l " label_type"
] [
.BI -M " hsmfile"
] [ 
.BI -N " nread"
] [
.BI -o
] [
.BI -p " pool"
] [
.BI -q " file_sequence_number"
] [
.BI -S " tape_server"
] [
.BI -s " size"
] [
.BI -T
] [
.BI -U " fun"
] [
.BI -u " user"
] [
.BI -V " visual_identifier(s)"
] [ 
.BI -v " volume_serial_number(s)"
] [
.BI -X " xparm"
] [ 
.BI -z
] [ 
.BI --nowait
] [ 
.BI --rdonly
] [ 
.BI --silent
] [ 
.I files
] 
.SH DESCRIPTION
This command sends a request to a stage daemon which checks its catalog.
If the file has already been staged, the catalog is updated
(number of accesses, last access time) and if
.I file
is different from previous job, a symbolic link is created and the control
given back to the user.
If the file is not staged yet, pools have been configured and
.B -p NOPOOL
is not specified, the file is allocated in a pool of which the space is managed
by the stage daemon.
If the request gives tape parameters, the internal file name is in the form
.B vid.fseq.label
while if an external file name is specified with the option -I (disk stageing),
the internal file name is in the form
.BR xfile_basename.reqid .
If no pool has been configured or
.B -p NOPOOL
is specified, the tape is staged directly in the location given by
.I file
and no space management is done.
To do the actual copy, a
.B tpread
is issued.
When complete, if
.I file
is specified, a symbolic link pointing at the actual file is created with that
name.
.LP
.I file
may be a local path, the combination
.I machine:fullpath
(NFS syntax), a full path starting with the NFS_ROOT, i.e. /shift,
or absent.
.SH OPTIONS
.TP
.BI \-A " allocation_policy"
Disk space allocation policy may be:
.RS
.TP
.B immediate
The disk space is allocated when receiving the stage request.
This is the default.
.TP
.B deferred
The disk space is allocated only when needed, i.e. one file at a time.
In case of multifile stageing, space is allocated for the first file, the copy
from tape to disk is done, the unused space is released; the process is 
then repeated for every tape file requested and found on the tape.
This mode increases the lifetime of files in the pool and is very useful when
files are of unknown size or vary a lot in size.
.RE
.TP
.BI \-b " blocksize"
Specifies the maximum block size.
If not given and the tape is labelled, the block size is taken from the
header label when the field is not zero; otherwise:
for all drive types except Redwoods the default is 32760 if
.I lrecl
is not specified,
.I lrecl
otherwise. On Redwoods, the default is 262144 if
.I lrecl
is not specified or the largest multiple of
.I lrecl
below 262144.
.TP
.BI \-c " off"
Stages \fBall\fP the file sequences of a tape onto separate disk files, given that you specify the starting file sequence through option -q followed by the file sequence number itself, followed again by a '-' character. This option is useless if not used in a tape request context. For example: 'stagein -V AB1234 -c off -q 17-' will stage all the file sequences on tape AB1234 from file sequence 17, up to the last one, into as many separate disk files.
.TP
.BI \-C " charconv"
The data is converted from this character set when doing the copy.
.I charconv
may be
.B ascii
(no conversion is done) or
.BR ebcdic .
The option value may also be
.BR block .
In this case, the trailing blanks are stripped off and a newline is appended
to each record.
Option values may be combined; they are separated by comma in this case.
.br
For example:
.br
	-C ebcdic,block
.TP
.BI \-d " density"
.B 1600
or
.B 6250
for round (IBM 3420) tapes,
.B 38000
for standard IBM 3480 cartridges,
.B 38KC
for 3480 cartridges with compaction (IDRC),
.B 38KD
for double density IBM 3490 cartridges,
.B 38KDC
for 3490 cartridges with compaction (IDRC),
.BR 8200 ,
.B 8500
or
.B 8500C
for Exabytes,
.B DDS
or
.B DDSC
for DAT cartridges,
.BR 2G ,
.BR 6G ,
.BR 10G ,
.BR 10GC ,
.BR 20G ,
.BR 20GC ,
.B 35G
or
.B 35GC
for DLT cartridges,
.B 10G
or
.B 10GC
for IBM 3590 cartridges,
.BR 10G ,
.BR 10GC ,
.BR 25G ,
.BR 25GC ,
.B 50G
or
.B 50GC
for Redwood cartridges.
It must match the device group specified in TMS (if installed).
If this parameter is missing, the value is taken from TMS.
.TP
.BI \-E " error_action"
Specifies the action to be taken in case of tape read error. This action may be:
.RS
.TP
.B skip
unreadable blocks are skipped. For each, an error message indicates the error
type and the block number.
A maximum of 10 blocks may be skipped.
.TP
.B keep
the file up to the block where the error occurred is kept on disk. By default,
the disk file is removed.
.TP
.B ignoreeoi
Do not take two consecutive tapemarks as End Of Information.
This option is only valid for label type
.B nl
or
.B blp
and should be avoided on IBM 3420 drives (the reel could wind off).
.RE
.TP
.BI \-F " recfm"
Specifies the record format of the data on the tape. It may be one of the 
following:
.RS
.TP
.B U,\|f77
Blocks may be of unknown length. A single logical record on disk corresponds
to one tape block. The disk copy of the file is formatted according to the 
convention for sequential 
.SM FORTRAN 
unformatted I/O.
.TP
.B U
is an alias for
.BR U,\|f77 .
.TP
.B U,\|bin
Blocks may be of unknown length. The data is transferred as a byte stream.
This format is suitable for use with direct access Fortran unformatted I/O
or C I/O.
.TP
.B F
Each tape block is exactly
.IR blocksize
bytes long (See the \-b option) with the exception of the last block
which may be shorter.
If both
.I blocksize
and
.I lrecl
are provided, they should be equal.
.br
Data is copied from tape to disk as a continuous data stream.
This format is suitable for use with direct access Fortran unformatted I/O
or C I/O.
.TP
.B FB
This allows to group several records of
.I lrecl
bytes into tape blocks of
.I blksize
bytes.
.I blksize
must be a multiple of
.IR lrecl .
.TP
.B F,\|-f77
The tape file is known to be formatted according to the convention 
for sequential 
.SM FORTRAN 
unformatted I/O. The control words are stripped off so that the disk file
is suitable for use with direct access 
.SM FORTRAN
unformatted I/O or C I/O.
The logical record length on the tape file must correspond to the record 
length specified either with the -L option or, if the -L option is not 
specified, the record length given in the label.
.LP
If the tape is labelled, the record format is taken from the header label;
otherwise the default is
.BR U .
.RE
.TP
.BI \-f " fid"
Specifies the fileid in HDR1. If more than 17 characters long,
only the last 17 are taken. It is then converted to upper case.
.I fid
is checked against the fid in HDR1.
In the case of multiple files, a series of file identifiers may be given,
separated by ":"s. These apply to the tape files in the specified order.
If there are more files than identifiers, the last one will be used for all
remaining files. Thus,
a single file identifier may be given which applies to all files on the tape.
.TP
.B \-G
Specifies that the tape copy operations should be issued on the tape server by
the 'group user'. A 'group user' may be defined for each group in 
.B /etc/shift.conf.
.br
For example: 
.br
	GRPUSER ws	opalprod
.TP 
.BI \-g " dgn"
Requests a device from an installation defined set of "device groups".
It must match the device group specified in TMS (if installed).
If this parameter is missing, the value is taken from TMS or
defaults to CART.
.TP
.BI \-h " host"
Sends the request to the stager on
.IR host .
This may also be specified thru the environment variable STAGE_HOST.
.TP
.BI \-I " external_filename"
Gives the name of an external disk file. In this case,
.B rfcp
is used to copy this file to the staging pool.
.TP
.BI \-L " lrecl"
Specifies the logical record length for fixed length record format files.
.TP
.BI \-l " lab_type"
Specifies the type of label which should be checked or generated. The following
label types are supported:
.RS
.TP
.B al
.SM ANSI 
label
.TP
.B blp
bypass label processing
.TP
.B nl
no label
.TP
.B sl
standard 
.SM IBM
label
.LP
It must match the label type specified in TMS (if installed).
If this parameter is missing, the value is taken from TMS or
defaults to
.BR sl .
.RE
.TP
.BI \-M " hsmfile"
Gives the name of a file stored in an Hierarchical Storage Manager (HSM).
.B rfcp
is used to copy the file to the staging pool. The option
.B -s
is not needed for a complete file stageing: the stager queries the HSM system
about the file size and allocates correctly the disk space in the pool.
Partial stageing is supported.
.TP
.BI \-N " nread"
If specified, only 
.I nread 
records will be copied.
In the case of multiple files, a series of nread values may be given,
separated by ":"s. These apply to the tape files in the specified order.
If there are more files than nread values, the last one will be used for all
remaining files. Thus,
a single nread value may be given which applies to all files on the tape.
.TP
.BI \-o
Implies old fileid. The existing fileid is checked if the option
.B -f
is also specified.
.TP
.BI \-p " poolname"
Allocates the file in the pool
.IR poolname .
This may also be specified thru the environment variable STAGE_POOL.
.TP
.BI \-q " fseq"
Specifies the file sequence number(s) of the tape file(s) to be copied or
created. Default is 1.
Multiple files may be specified using the following syntax:
.RS 
.TP 1.5i 
.I f1\-f2 
Files 
.I f1
to 
.I f2 
inclusive.
.TP
.I f1\- 
Files 
.I f1
to the last file on the tape.
.TP
.I f1\-f2,\|f4,\|f6- 
A series of non\-consecutive ranges of files.
.TP
.B u
Position the tape by fid.
.TP
.BI u x
Position the tape by fid on
.I x
different files.  A series of
.I x
fileids must be given as
.B -f
option value.
.RE
.TP
.BI \-S " server"
Specifies that the tape copy operation has to be issued on the tape server 
.IR server. 
.TP
.BI \-s " size"
If specified, only 
.I size 
Mbytes will be copied.
The value must be smaller than 2048 (2 Gbytes).
In the case of multiple files, a series of size values may be given,
separated by ":"s. These apply to the tape files in the specified order.
If there are more files than size values, the last one will be used for all
remaining files. Thus,
a single size value may be given which applies to all files on the tape.
.TP
.BI \-T
Does not check trailer labels.
.TP
.BI \-U " fun"
Fortran unit number.
This is internally converted to a local path
.BI ftn fun
(HP-UX) or
.BI fort. fun
(all other platforms).
The
.B \-U
option and the
.I file
parameter may be both specified for single file stageing only.
.TP
.BI \-u " user"
A file allocated in a pool will be stored in a subdirectory
.IR user .
This may also be specified thru the environment variable STAGE_USER.
The default subdirectory name is
.BR stage .
.I user
must be a valid user name and belong to the same group as the requestor.
.TP
.BI \-V " vid"
Specifies the tape 
.SM VID
if different from the 
.SM VSN .
In the case of multi\-volume files a series of 
.IR vid s
may be given, separated by ":"s, in the order in which the tapes are to be used.
.TP
.BI \-v " vsn"
Specifies the volume serial number of the tape. In the case of a multi-volume
file a series of 
.IR vsn s
may be given, separated by ":"s, in the order in which the tapes are to be used.
When \-v and \-V are both used, each
.IR vsn
must be matched by a
.IR vid
in the same order.
.TP
.B \-z
Gives on standard error, as an option string, the main characteristics of the 
tape file just staged. This includes block size (-b), record format (-F), file
identifier (-f), record length (-L), full pathname of diskfile in the pool (-P)
and file sequence number (-q).
The line is prefixed by the string "STG47 - ". This flag is set by
.BR copytape .
For non-tape requests, this prints only the full pathname.
.TP
.B \--nowait
Does not hang the client. Recall is done internally without further notification.
.TP
.B \--rdonly
Tells the stager that you are requesting this file for read-only purpose.
.TP
.B \--silent
Tells the stager not to output you any message
.TP
.SH EXAMPLES
.LP
To get the data from a tape into the default stage pool:
.RS
stagein -v CIN111 inp
.RE
.SH RETURN CODES
\
.br
0	Ok.
.br
1	Bad parameter.
.br
2	System error.
.br
4	Configuration error.
.br
16	EBUSY: Device or resource busy.
.br
28	ENOSPC: No space left on device.
.br
188	Tape hosting the data is held
.br
192	Request cancelled by stageclr.
.br
193	Blocks were skipped.
.br
194	Blocks were skipped and stageing limited by size.
.br
195	Stagein stopped: too many tape errors, but -E keep.
.br
196	Request killed
.br
197	Stageing limited by size.
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


castor/client/StageIn.cpp
/******************************************************************************
 *                      StageIn.cpp
 *
 * This file is part of the Castor project.
 * See http://castor.web.cern.ch/castor
 *
 * Copyright (C) 2003  CERN
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * @(#)$RCSfile: StageIn.cpp,v $ $Revision: 1.7 $ $Release$ $Date: 2004/07/13 13:36:28 $ $Author: sponcec3 $
 *
 *
 *
 * @author Sebastien Ponce
 *****************************************************************************/

// Include Files
#include <iostream>
#include <cstdlib>
#include <vector>
#include "castor/rh/File.hpp"
#include "castor/rh/StageInRequest.hpp"
#include "castor/exception/Exception.hpp"
#include "stage_constants.h"
#include "common.h"
#include "stage_api.h"

// Local Files
#include "StageIn.hpp"

//------------------------------------------------------------------------------
// buildRequest
//------------------------------------------------------------------------------
castor::rh::Request* castor::client::StageIn::buildRequest()
  throw (castor::exception::Exception) {
  u_signed64 flags = 0;
  int openflags = 0;
  {
    // RH server host. Can be passed in the -h option
    // or given through the RH_HOST environment variable
    // or given in the castor.conf file as a RH/HOST entry
    char* host;
    if (m_inputFlags.find("h") != m_inputFlags.end()) {
      m_rhHost = m_inputFlags["h"];
    } else if ((host = getenv ("RH_HOST")) != 0 ||
               (host = getconfent("RH","HOST",0)) != 0) {
      m_rhHost = host;
    } else {
      castor::exception::Exception e(ETPRM);
      e.getMessage()
        << "Unable to deduce the name of the RH server.\n"
        << "No -h option was given, RH_HOST is not set and "
        << "your castor.conf file does not contain a RH/HOST entry."
        << std::endl;
      throw e;
    }
  }
  {
    char* port;
    // RH server port. Can be given through the environment
    // variable RH_PORT or in the castor.conf file as a
    // RH/PORT entry
    if ((port = getenv ("RH_PORT")) != 0 ||
        (port = getconfent("RH","PORT",0)) != 0) {
      int iport;
      char* dp;
      if (stage_strtoi(&iport, port, &dp, 0) != 0) {
        castor::exception::Exception e(errno);
        e.getMessage() << "Bad port value." << std::endl;
        throw e;
      }
      if (iport < 0) {
        castor::exception::Exception e(errno);
        e.getMessage()
          << "Invalid port value : " << iport
          << ". Must be > 0." << std::endl;
        throw e;        
      }
      m_rhPort = iport;
    } else {
      castor::exception::Exception e(ETPRM);
      e.getMessage()
        << "Unable to deduce the RH server port.\n"
        << "RH_PORT is not set and your castor.conf file "
        << "does not contain a RH/PORT entry."
        << std::endl;
      throw e;
    }
  }
  std::string poolName;
  {
    char* pool;
    // Pool name. Can be given by the -p option
    // or through the environment variable STAGE_POOL
    if (m_inputFlags.find("p") != m_inputFlags.end()) {
      poolName = m_inputFlags["p"];
    } else if ((pool = getenv ("STAGE_POOL")) != 0) {
      poolName = pool;
    } else {
      castor::exception::Exception e(ETPRM);
      e.getMessage()
        << "Unable to deduce the name of the stage pool.\n"
        << "No -p option was given and STAGE_POOL is not set."
        << std::endl;
      throw e;
    }
  }
  // -K option (parsed by BaseClient)
  if (m_inputFlags.find("K") != m_inputFlags.end()) {
    castor::exception::Exception e(ETPRM);
    e.getMessage()
      << "-K option is not valid in the stagein command."
      << std::endl;
    throw e;    
  }
  // Allocation policy
  if (m_inputFlags.find("A") != m_inputFlags.end()) {
    if (m_inputFlags["A"] == "deferred") {
      flags |= STAGE_DEFERRED;
    } else if (m_inputFlags["A"] != "immediate") {
      castor::exception::Exception e(ETPRM);
      e.getMessage()
        << "Invalid argument for -A option.\n"
        << "Should be either deferred or immediate."
        << std::endl;
      throw e;
    }
  }
  // Size
  std::vector<u_signed64> sizes;
  if (m_inputFlags.find("s") != m_inputFlags.end()) {
    std::string ssize = m_inputFlags["s"];
    char* size = (char*) malloc(ssize.length()+1);
    strncpy(size, ssize.c_str(), ssize.length());
    size[ssize.length()] = 0;
    char* p = strtok (size, ":");
    while (p != 0) {
      u_signed64 size;
      if (stage_util_check_for_strutou64(p) < 0 ||
          (size = strutou64(p)) <= 0) {
        castor::exception::Exception e(ETPRM);
        e.getMessage()
          << "Invalid syntax in -s option."
          << std::endl;
        throw e;
      }
      sizes.push_back(size);
      p = strtok(0, ":");
    }
    free(size);
  }
  // Silent
  if (m_inputFlags.find("silent") != m_inputFlags.end()) {
    flags |= STAGE_SILENT;
  }
  // NoWait
  if (m_inputFlags.find("nowait") != m_inputFlags.end()) {
    flags |= STAGE_NOWAIT;
  }
  // NoRetry
  if (m_inputFlags.find("noretry") != m_inputFlags.end()) {
    flags |= STAGE_NORETRY;
  }
  // RdOnly
  if (m_inputFlags.find("rdonly") != m_inputFlags.end()) {
    openflags |= STAGE_RDONLY;
  }
  // Build request
  castor::rh::StageInRequest* req =
    new castor::rh::StageInRequest();
  req->setFlags(flags);
  req->setOpenflags(openflags);
  int n = 0;
  for (std::vector<std::string>::const_iterator it = m_inputArguments.begin();
       it != m_inputArguments.end();
       it++) {
    castor::rh::File* f = new castor::rh::File();
    f->setName(*it);
    f->setPoolname(poolName);
    if (n < sizes.size()) {
      f->setXsize(sizes[n]);
    } else {
      f->setXsize(0);
    }
    n++;
    req->addFiles(f);
    f->setRequest(req);
  }
  return req;
}

//------------------------------------------------------------------------------
// usage
//------------------------------------------------------------------------------
void castor::client::StageIn::usage(std::string error) throw() {
  std::cout << error << std::endl;
  std::cout << "usage : stagein [-A alloc_mode] [-h rh_host] [-p pool] [-s size] "
            << "[--noretry] [--nowait] [--silent] [--rdonly]"
            << " hsmfile..." << std::endl;
}
